src/Chainweb/Pact/Service/PactQueue.hs

```diff
@@ -12,37 +18,79 @@
 module Chainweb.Pact.Service.PactQueue
     ( addRequest
     , getNextRequest
+    , getPactQueueStats
     , newPactQueue
+    , resetPactQueueStats
     , PactQueue
     ) where
 
+import Data.Aeson
 import Control.Applicative
 import Control.Concurrent.STM.TBQueue
+import Control.DeepSeq (NFData)
+import Control.Lens (each)
 import Control.Monad ((>=>))
 import Control.Monad.STM
+import Data.IORef
+import qualified Data.Text as T
+import Data.Tuple.Strict
+import GHC.Generics
 import Numeric.Natural
-
 import Chainweb.Pact.Service.Types
+import Chainweb.Time
 
 -- | The type of the Pact Queue
 -- type PactQueue = TBQueue RequestMsg
 data PactQueue = PactQueue
   {
-    _pactQueueValidateBlock :: !(TBQueue RequestMsg)
-  , _pactQueueNewBlock :: !(TBQueue RequestMsg)
-  , _pactQueueOtherMsg :: !(TBQueue RequestMsg)
+    _pactQueueValidateBlock :: !(TBQueue (T2 RequestMsg (Time Micros)))
+  , _pactQueueNewBlock :: !(TBQueue (T2 RequestMsg (Time Micros)))
+  , _pactQueueOtherMsg :: !(TBQueue (T2 RequestMsg (Time Micros)))
+  , _pactQueuePactQueueValidateBlockMsgStats :: !PactQueueStats
+  , _pactQueuePactQueueNewBlockMsgStats :: !PactQueueStats
+  , _pactQueuePactQueueOtherMsgStats :: !PactQueueStats
   }
 
-newPactQueue :: Natural -> STM PactQueue
+initPactQueueCounters :: PactQueueCounters
+initPactQueueCounters = PactQueueCounters
+    {
+      _pactQueueCountersCount = 0
+    , _pactQueueCountersSum = 0
+    , _pactQueueCountersMin = maxBound
+    , _pactQueueCountersMax = 0
+    }
+
+newPactQueue :: Natural -> IO PactQueue
 newPactQueue sz = do
-  _pactQueueValidateBlock <- newTBQueue sz
-  _pactQueueNewBlock <- newTBQueue sz
-  _pactQueueOtherMsg <- newTBQueue sz
+  (_pactQueueValidateBlock, _pactQueueNewBlock, _pactQueueOtherMsg) <- each newTBQueueIO (sz,sz,sz)
+  _pactQueuePactQueueValidateBlockMsgStats <- do
+      counters <- newIORef initPactQueueCounters
+      return PactQueueStats
+        {
+          _pactQueueStatsQueueName = "ValidateBlockMsg"
+        , _pactQueueStatsCounters = counters
+        }
+  _pactQueuePactQueueNewBlockMsgStats <- do
+      counters <- newIORef initPactQueueCounters
+      return PactQueueStats
+        {
+          _pactQueueStatsQueueName = "NewBlockMsg"
+        , _pactQueueStatsCounters = counters
+        }
+  _pactQueuePactQueueOtherMsgStats <- do
+      counters <- newIORef initPactQueueCounters
+      return PactQueueStats
+        {
+          _pactQueueStatsQueueName = "OtherMsg"
+        , _pactQueueStatsCounters = counters
+        }
   return PactQueue {..}
```

Comment:
Please avoid `RecordWildCards`. If you want to avoid an helper variable, you may consider applicative functor notation (`<$>` and `<*>`). Otherwise just being explicit is fine, too.